### PR TITLE
added task 40, singleton driver pattern, updated Page Object pattern

### PR DESCRIPTION
**Pull request by KonstantinLevchukGree**
A lot of changes

### CODE REVIEW

File: src/test/java/seleniumeasyPage/ProgressBarPage.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        int percentProgressBar= 0;
+        while (Integer.parseInt(driver.findElement(PERCENT_PROGRESS_BAR).getText().replace("%","").trim()) <= Integer.parseInt(dataTests.getProperty("percentProgressBar"))) {
+
+            percentProgressBar=Integer.parseInt(driver.findElement(PERCENT_PROGRESS_BAR).getText().replace("%","").trim());
+
+            if(percentProgressBar>Integer.parseInt(dataTests.getProperty("percentProgressBar"))){
+                driver.navigate().refresh();
+                break;
+            }
+        }
```
Comment (KirKarpov) [inline src/test/java/seleniumeasyPage/ProgressBarPage.java]: There is a really good expected condition that I would like to try: ExpectedConditions.textMatches() 

https://www.selenium.dev/selenium/docs/api/java/org/openqa/selenium/support/ui/ExpectedConditions.html#textMatches(org.openqa.selenium.By,java.util.regex.Pattern)

File: src/test/java/seleniumeasyTests/ProgressBarTest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ProgressBarTest extends BaseTest {
+    private ProgressBarPage progressBarPage;
+    private final Properties dataTests = PropertyUtil.getProperties("testsData.properties");
+
+    @Test
+    public void refreshProgressBar() {
+        progressBarPage = new ProgressBarPage(LocalChromeDriver.getInstance());
+        assertTrue(progressBarPage.getPercentProgressBar() >= Integer.parseInt(dataTests.getProperty("percentProgressBar")), "ProgressBarPage not refreshed");
```
Comment (KirKarpov) [inline src/test/java/seleniumeasyTests/ProgressBarTest.java]: You need to check that page was refreshed after you call getRercentProgressBar()

File: src/test/java/seleniumeasyTests/TableSortSearchTest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+public class TableSortSearchTest extends BaseTest {
+    private TableSortSearchPage tableSortSearchPage;
+    private final Properties dataTests = PropertyUtil.getProperties("testsData.properties");
+
+    @Test
+    public void getUserByFilter() {
+
+        tableSortSearchPage = new TableSortSearchPage(LocalChromeDriver.getInstance());
+
+        List<User> listUsers = tableSortSearchPage.getUsersByAgeAndSalary(Integer.parseInt(dataTests.getProperty("minUserAge"))
```
Comment (KirKarpov) [inline src/test/java/seleniumeasyTests/TableSortSearchTest.java]: If you are reading data from property file, you can do it in the method. No need to do it on test level. So you method will be without parameters

File: src/test/java/mailYandexTests/ParametrizedLoginTest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+public class ParametrizedLoginTest extends BaseTest {
+    private StartPage startPage;
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/userData.csv")
+
+    public void authentication(String userLogin, String UserPassword) {
+
+        startPage = new StartPage(LocalChromeDriver.getInstance());
+
```
Comment (KirKarpov) [inline src/test/java/mailYandexTests/ParametrizedLoginTest.java]: Be careful with spaces between methods. Sometimes they aren't needed

File: src/test/java/seleniumeasyTests/AlertBoxTest.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import seleniumeasyPage.AlertPage;
+import utils.singleton.LocalChromeDriver;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class AlertBoxTest extends BaseTest {
+    private AlertPage alertPage;
+
+    @Test
+    public void openAlert() {
```
Comment (KirKarpov) [inline src/test/java/seleniumeasyTests/AlertBoxTest.java]: Test name should  say what it check. Usually it starts from the words: verify, check, test. As example the name could  verifyAlertAcceptMessage() 
Please check naming for other methods in classes with tests

File: src/test/resources/testsData.properties [hunk 6]
```diff
@@ -1,3 +1,15 @@
 userName=konstantinliauchuk
```
Comment (KirKarpov) [inline src/test/resources/testsData.properties]: Name should be divided by dot. Example user.name= konstantinliauchuk

File: src/test/resources/testsData.properties [hunk 7]
```diff
@@ -1,3 +1,15 @@
...
 userPassword=11436811a
-uri=https://mail.yandex.com/
\ No newline at end of file
+mailYandexUrl=https://mail.yandex.com/
+seleniumEasySelectUrl=https://demo.seleniumeasy.com/basic-select-dropdown-demo.html
+seleniumEasyAlertUrl=https://demo.seleniumeasy.com/javascript-alert-box-demo.html
+seleniumUserUrl=https://demo.seleniumeasy.com/dynamic-data-loading-demo.html
```
Comment (KirKarpov) [inline src/test/resources/testsData.properties]: it is better to make name shorter. for example : select.dropdown.url, js.alert.url, dynamic.loading.url and etc.

**Review by KirKarpov — CHANGES_REQUESTED** [review]

File: src/test/java/mailYandexTests/ParametrizedLoginTest.java (hunk 4 above)
Comment (KonstantinLevchukGree) [reply src/test/java/mailYandexTests/ParametrizedLoginTest.java]: fixed

File: src/test/java/seleniumeasyTests/AlertBoxTest.java (hunk 5 above)
Comment (KonstantinLevchukGree) [reply src/test/java/seleniumeasyTests/AlertBoxTest.java]: fixed

File: src/test/resources/testsData.properties (hunk 6 above)
Comment (KonstantinLevchukGree) [reply src/test/resources/testsData.properties]: fixed

File: src/test/resources/testsData.properties (hunk 7 above)
Comment (KonstantinLevchukGree) [reply src/test/resources/testsData.properties]: fixed

File: src/test/java/seleniumeasyTests/TableSortSearchTest.java (hunk 3 above)
Comment (KonstantinLevchukGree) [reply src/test/java/seleniumeasyTests/TableSortSearchTest.java]: fixed

File: src/test/java/seleniumeasyPage/ProgressBarPage.java (hunk 1 above)
Comment (KonstantinLevchukGree) [reply src/test/java/seleniumeasyPage/ProgressBarPage.java]: fixed

File: src/test/java/seleniumeasyTests/ProgressBarTest.java (hunk 2 above)
Comment (KonstantinLevchukGree) [reply src/test/java/seleniumeasyTests/ProgressBarTest.java]: fixed

**Review by KirKarpov — APPROVED** [review]

**Pull request merged**